File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
/*

입력
첫 번째 줄에 수빈이가 있는 위치 N과 동생이 있는 위치 K가 주어진다. N과 K는 정수이다.

출력
수빈이가 동생을 찾는 가장 빠른 시간을 출력한다.

bfs

*/



import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

    public static int bfs(int N, int K) {
        int maxLimit = 100001;
        int[] visited = new int[maxLimit]; // 방문 시간 저장 배열

        // 큐를 초기화하고 시작점 N을 넣음
        Queue<Integer> queue = new LinkedList<>();
        queue.add(N);
        visited[N] = 1; // 시작점을 방문했음을 표시 (시간을 1로 설정)

        while (!queue.isEmpty()) {
            int currentPosition = queue.poll();
            
            // 동생을 찾았을 경우
            if (currentPosition == K) {
                return visited[currentPosition] - 1; // 시작점을 1로 했으므로 결과에서 1을 뺌
            }

            // 다음으로 이동할 세 가지 위치를 탐색
            int[] nextPositions = {currentPosition - 1, currentPosition + 1, currentPosition * 2};
            for (int nextPosition : nextPositions) {
                if (nextPosition >= 0 && nextPosition < maxLimit && visited[nextPosition] == 0) {
                    queue.add(nextPosition);
                    visited[nextPosition] = visited[currentPosition] + 1; // 이전 위치에서 1초 추가
                }
            }
        }
        return -1; // 이 경우는 발생하지 않음
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 입력 받기
        String[] input = br.readLine().split(" ");
        int N = Integer.parseInt(input[0]);
        int K = Integer.parseInt(input[1]);

        // BFS 탐색 결과 출력
        System.out.println(bfs(N, K));
    }
}

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque

def find_fastest_time(N, K):
    # 최대 위치 설정
    max_position = 100000
    # 방문 여부를 체크하기 위한 리스트
    visited = [False] * (max_position + 1)
    
    # BFS를 위한 큐 설정
    queue = deque([(N, 0)])  # (현재 위치, 현재 시간)
    visited[N] = True
    
    while queue:
        current_position, current_time = queue.popleft()
        
        # 동생을 찾았을 경우 현재 시간을 반환
        if current_position == K:
            return current_time
        
        # 다음 가능한 이동을 탐색
        next_positions = [current_position - 1, current_position + 1, current_position * 2]
        
        for next_position in next_positions:
            if 0 <= next_position <= max_position and not visited[next_position]:
                visited[next_position] = True
                queue.append((next_position, current_time + 1))

# 입력
N, K = map(int, input().split())

# 결과 출력
print(find_fastest_time(N, K))
